File: src/materials.cpp
#include "../include/geekatplay/materials.hpp"
#include <cmath>
#include <algorithm>

namespace Geekatplay {

std::vector<float> MaterialSolver::ComputeSlopeMap(const std::vector<float>& inH, int res) {
    std::vector<float> slope(res * res, 0.0f);
    for (int y = 1; y < res - 1; ++y) {
        for (int x = 1; x < res - 1; ++x) {
            float dx = (inH[y * res + (x + 1)] - inH[y * res + (x - 1)]) * 0.5f;
            float dy = (inH[(y + 1) * res + x] - inH[(y - 1) * res + x]) * 0.5f;
            float s = std::atan(std::sqrt(dx * dx + dy * dy)) / (3.1415926535f * 0.5f);
            slope[y * res + x] = std::clamp(s, 0.0f, 1.0f);
        }
    }
    return slope;
}

std::vector<float> MaterialSolver::ComputeCurvatureMap(const std::vector<float>& inH, int res) {
    std::vector<float> curv(res * res, 0.5f);
    for (int y = 1; y < res - 1; ++y) {
        for (int x = 1; x < res - 1; ++x) {
            float h = inH[y * res + x];
            float lap = inH[(y - 1) * res + x] + inH[(y + 1) * res + x] + inH[y * res + (x - 1)] + inH[y * res + (x + 1)] - 4.0f * h;
            curv[y * res + x] = std::clamp(lap * 10.0f + 0.5f, 0.0f, 1.0f);
        }
    }
    return curv;
}

std::vector<float> MaterialSolver::ComputeNormalMap(const std::vector<float>& inH, int res, float strength) {
    std::vector<float> normalRGB(res * res * 3, 0.0f);
    for (int y = 1; y < res - 1; ++y) {
        for (int x = 1; x < res - 1; ++x) {
            float dx = (inH[y * res + (x + 1)] - inH[y * res + (x - 1)]) * 0.5f;
            float dy = (inH[(y + 1) * res + x] - inH[(y - 1) * res + x]) * 0.5f;

            float nx = -dx * strength;
            float ny = -dy * strength;
            float nz = 1.0f;

            float len = std::sqrt(nx * nx + ny * ny + nz * nz) + 1e-6f;
            nx /= len; ny /= len; nz /= len;

            int idx = (y * res + x) * 3;
            normalRGB[idx + 0] = (nx + 1.0f) * 0.5f;
            normalRGB[idx + 1] = (ny + 1.0f) * 0.5f;
            normalRGB[idx + 2] = (nz + 1.0f) * 0.5f;
        }
    }
    return normalRGB;
}

PBRSplatmapResult MaterialSolver::GeneratePBRSplatmap(
    const std::vector<float>& inH, int res, const std::vector<float>& flowMap, float snowLine, float cliffSlopeThreshold
) {
    PBRSplatmapResult result;
    result.rockMask.assign(res * res, 0.0f);
    result.snowMask.assign(res * res, 0.0f);
    result.soilMask.assign(res * res, 0.0f);
    result.gravelMask.assign(res * res, 0.0f);
    result.splatmapRGBA.assign(res * res * 4, 0.0f);

    auto slope = ComputeSlopeMap(inH, res);
    result.normalMapRGB = ComputeNormalMap(inH, res);

    for (int i = 0; i < res * res; ++i) {
        float h = inH[i];
        float s = slope[i];

        // 1. Rock Mask
        float rock = std::clamp((s - cliffSlopeThreshold * 0.7f) / (cliffSlopeThreshold * 0.6f + 1e-6f), 0.0f, 1.0f);

        // 2. Snow Mask
        float snow = std::clamp((h - snowLine) / 0.18f, 0.0f, 1.0f) * (1.0f - rock * 0.6f);

        float flowVal = (i < static_cast<int>(flowMap.size())) ? flowMap[i] : 0.0f;

        // 3. Soil / Vegetation (Enhanced by flow wetness)
        float soil = (1.0f - rock) * (1.0f - snow) * std::clamp(1.0f - h * 1.2f + flowVal * 0.2f, 0.0f, 1.0f);

        // 4. Gravel / Scree
        float gravel = std::clamp(1.0f - (rock + snow + soil), 0.0f, 1.0f);

        result.rockMask[i] = rock;
        result.snowMask[i] = snow;
        result.soilMask[i] = soil;
        result.gravelMask[i] = gravel;

        int rgbaIdx = i * 4;
        result.splatmapRGBA[rgbaIdx + 0] = rock;
        result.splatmapRGBA[rgbaIdx + 1] = gravel;
        result.splatmapRGBA[rgbaIdx + 2] = soil;
        result.splatmapRGBA[rgbaIdx + 3] = snow;
    }

    return result;
}

} // namespace Geekatplay

File: core/materials.py
"""
Geekatplay Studio — PBR Materials, Normal Baking & Biome Splatmaps.
Computes slope, curvature, flow masks, triplanar blending, and high-fidelity PBR splatmaps.
"""

import numpy as np
from typing import Dict, Tuple, Optional


def compute_slope_map(heightfield: np.ndarray) -> np.ndarray:
    """Computes terrain surface slope angles in normalized [0, 1] range (0 to 90 degrees)."""
    h = heightfield.astype(np.float32)
    dy, dx = np.gradient(h)
    slope = np.arctan(np.sqrt(dx * dx + dy * dy)) / (np.pi * 0.5)
    return np.clip(slope, 0.0, 1.0).astype(np.float32)


def compute_curvature_map(heightfield: np.ndarray) -> np.ndarray:
    """Computes Laplacian surface curvature (ridge vs valley / cavity)."""
    h = heightfield.astype(np.float32)
    laplacian = (
        np.pad(h, ((1, 0), (0, 0)), mode="edge")[:-1, :] +
        np.pad(h, ((0, 1), (0, 0)), mode="edge")[1:, :] +
        np.pad(h, ((0, 0), (1, 0)), mode="edge")[:, :-1] +
        np.pad(h, ((0, 0), (0, 1)), mode="edge")[:, 1:] -
        4.0 * h
    )
    # Normalize centered at 0.5
    norm_curv = np.clip(laplacian * 10.0 + 0.5, 0.0, 1.0)
    return norm_curv.astype(np.float32)


def compute_normal_map(heightfield: np.ndarray, strength: float = 8.0) -> np.ndarray:
    """
    Computes tangent-space normal map RGB [0, 1] from heightfield gradients.
    """
    h = heightfield.astype(np.float32)
    dy, dx = np.gradient(h)

    # Tangent space normal vectors: (-dx * strength, -dy * strength, 1.0)
    nx = -dx * strength
    ny = -dy * strength
    nz = np.ones_like(h)

    length = np.sqrt(nx * nx + ny * ny + nz * nz) + 1e-6
    nx /= length
    ny /= length
    nz /= length

    # Remap from [-1, 1] to [0, 1] RGB
    r = (nx + 1.0) * 0.5
    g = (ny + 1.0) * 0.5
    b = (nz + 1.0) * 0.5

    return np.stack([r, g, b], axis=-1).astype(np.float32)


def generate_pbr_splatmap(
    heightfield: np.ndarray,
    flow_map: Optional[np.ndarray] = None,
    snow_line: float = 0.72,
    cliff_slope_threshold: float = 0.45,
) -> Dict[str, np.ndarray]:
    """
    Generates multi-channel PBR biome splatmaps:
    - Channel 0 (Red): Steep Rock & Granite Cliffs
    - Channel 1 (Green): Talus Scree & Valley Gravel
    - Channel 2 (Blue): Lowland Grass & Humus Soil
    - Channel 3 (Alpha): High-altitude Alpine Snow & Ice
    - Wetness / River Flow channel
    """
    h = heightfield.astype(np.float32)
    slope = compute_slope_map(h)
    curv = compute_curvature_map(h)
    flow = flow_map if flow_map is not None else np.zeros_like(h)

    # 1. Rock Mask (Steep slopes)
    rock = np.clip((slope - cliff_slope_threshold * 0.7) / (cliff_slope_threshold * 0.6 + 1e-6), 0.0, 1.0)

    # 2. Snow Mask (High elevation, gathers in low slope depressions)
    snow = np.clip((h - snow_line) / 0.18, 0.0, 1.0) * (1.0 - rock * 0.6)

    # 3. Wetness / River Mask
    wetness = np.clip(flow * 1.5, 0.0, 1.0)

    # 4. Soil / Vegetation Mask (Low slope, below snow line)
    soil = (1.0 - rock) * (1.0 - snow) * np.clip(1.0 - h * 1.2, 0.0, 1.0)

    # 5. Scree / Gravel Mask (Remaining transition areas)
    gravel = np.clip(1.0 - (rock + snow + soil), 0.0, 1.0)

    splat_rgba = np.stack([rock, gravel, soil, snow], axis=-1)

    return {
        "splatmap_rgba": splat_rgba.astype(np.float32),
        "rock_mask": rock.astype(np.float32),
        "snow_mask": snow.astype(np.float32),
        "soil_mask": soil.astype(np.float32),
        "gravel_mask": gravel.astype(np.float32),
        "wetness_mask": wetness.astype(np.float32),
    }
